PSA C configuration: more concrete information (#3697)
ARMmbed/mbedtls: docs/proposed/psa-conditional-inclusion-c.md

This is currently a proposal for Mbed TLS. It is not currently on track for standardization in PSA.

> Time-stamp: "2020/09/21 18:07:09 GMT"

Author: > Does it mean that if a PSA driver provides a crypto mechanism then "automatically" the Mbed TLS code providing the mechanism is removed from the Mbed TLS build? If yes, what about not to do that, at least to start with, to simplify.

The short-term goal is to make it _possible_ to include e.g. PSA RSA support without including `MBEDTLS_RSA_C` when there's a PSA transparent driver for RSA. I don't see a simplification that would make it possible but not automatic. At least not without making a temporary mechanism that we would not want to maintain backward compatibility with because it would be too ad hoc and hard to use.

> Why not just tightening mbedtls/config.h to psa/config.h in that case: the defines in psa/config.h define the defines in mbedtls/config.h?

I don't understand what you're proposing. Is this about the extra layer of indirection with `MBEDTLS_HAVE_BUILTIN_xxx` being aliases for `MBEDTLS_xxx_C`? That requires more work in `config_psa.h` but less in `psa_crypto.c` and other places. It makes the code more readable because there isn't always a 1-to-1 correspondence between `MBEDTLS_HAVE_BUILTIN_xxx` and the legacy symbols, and this the defining expression is only present once.

> Time-stamp: "2020/09/21 18:07:09 GMT"

Reviewer: > The short-term goal is to make it possible to include e.g. PSA RSA support without including MBEDTLS_RSA_C when there's a PSA transparent driver for RSA. I don't see a simplification that would make it possible but not automatic. At least not without making a temporary mechanism that we would not want to maintain backward compatibility with because it would be too ad hoc and hard to use.

My understanding of the current proposal is that, the integrator defines in psa/crypto_config.h the needs of the application, defines the drivers and their capabilities and from that the build system computes the needs for built-in crypto.

What I am suggesting is rather for the integrator to defines in psa/crypto_config.h the needs in built-in crypto. Thus the integrator has to determine herself the needs for built-in crypto given the application needs and the drivers capabilities.

To follow-up on the RSA example. if currently we disable MBEDTLS_RSA_C in config.h what do have to change in psa_crypto.c for the driver wrapper to be called for RSA operations ?

> Time-stamp: "2020/09/21 18:07:09 GMT"

Author: Requiring that the integrator define the needs for built-in crypto requires the integrator to know the needs of the application and to understand the MBEDTLS configuration symbols. This puts a significant burden on the integrator which can be avoided by a small amount of code: we can write the rules to deduce the MBEDTLS symbols from the PSA ones once and for all. We need to do the analysis to figure out the requirements in order to write the rules for the other direction (which are necessary for backward compatibility), so the cost for us to write the rules for the PSA→MBEDTLS is just the cost of writing those few extra lines.

For RSA, `defined(MBEDTLS_RSA_C)` in `psa_crypto.c` will change to one of, or a combination of, `MBEDTLS_PSA_BUILTIN_KEY_TYPE_RSA`, `MBEDTLS_PSA_BUILTIN_ALG_RSA_PKCS1V15_SIGN_RAW`, `MBEDTLS_PSA_BUILTIN_ALG_RSA_PSS`, `MBEDTLS_PSA_BUILTIN_ALG_RSA_PKCS1V15_CRYPT` or `MBEDTLS_PSA_BUILTIN_ALG_RSA_OAEP` depending on the location (PSA has more granularity there). `defined(MBEDTLS_PKCS1_V15)` will change to one of `MBEDTLS_PSA_BUILTIN_ALG_RSA_PKCS1V15_SIGN_RAW` or `MBEDTLS_PSA_BUILTIN_ALG_RSA_PKCS1V15_CRYPT`. `defined(MBEDTLS_PKCS1_V21)` will change to one of `MBEDTLS_PSA_BUILTIN_ALG_RSA_PSS`  or `MBEDTLS_PSA_BUILTIN_ALG_RSA_OAEP`. `defined(MBEDTLS_RSA_C) && defined(MBEDTLS_RSA_GENPRIME)` will change to will change to `MBEDTLS_PSA_BUILTIN_KEY_TYPE_RSA` (PSA has less granularity there).

> Time-stamp: "2020/09/21 18:07:09 GMT"

Reviewer: Yes I understand now that the conversion from PSA symbols to MBEDTLS symbols is not a big work and as it eases integration it makes sense to do it.

Otherwise regarding:

>> Why not just tightening mbedtls/config.h to psa/config.h in that case: the defines in psa/config.h define the defines in mbedtls/config.h?

> I don't understand what you're proposing. Is this about the extra layer of indirection with MBEDTLS_HAVE_BUILTIN_xxx being aliases for MBEDTLS_xxx_C? That requires more work in config_psa.h but less in psa_crypto.c and other places. It makes the code more readable because there isn't always a 1-to-1 correspondence between MBEDTLS_HAVE_BUILTIN_xxx and the legacy symbols, and this the defining expression is only present once.

I am not sure why when `MBEDTLS_PSA_CRYPTO_CONFIG` is set the MBEDTLS symbols are not just deduced from the PSA ones. Following-up on the AES example in the "PSA Crypto configuration file" section, why do we give the possibility of defining MBEDTLS_AES_C if given psa/crypto_config.h there is no need for it?

> Time-stamp: "2020/09/21 18:07:09 GMT"

Author: When `MBEDTLS_PSA_CRYPTO_CONFIG` is set, only the `MBEDTLS_xxx` symbols that are needed for PSA calls are automatically enabled. The PSA crypto configuration can't determine whether to enable `MBEDTLS_CIPHER_C` or `MBEDTLS_SSL_CLI_C`.

It's `psa/crypto_config.h` and not `psa/config.h` because this configuration mechanism is specific to crypto. Other PSA components manage their configuration differently (likely via JSON).

I'd forgotten to make the editorial corrections from this comment thread, I just pushed them.

> Time-stamp: "2020/09/21 18:07:09 GMT"

Reviewer: > When `MBEDTLS_PSA_CRYPTO_CONFIG` is set, only the `MBEDTLS_xxx` symbols that are needed for PSA calls are automatically enabled. The PSA crypto configuration can't determine whether to enable `MBEDTLS_CIPHER_C` or `MBEDTLS_SSL_CLI_C`.

Coming back to this, I understand your point for `MBEDTLS_SSL_CLI_C` (TLS, not crypto configuration) but I don't really understand why we would need to enable `MBEDTLS_CIPHER_C` if it is not required by `PSA_WANT_xxx` macros. Do we want to support the case when MBEDTLS_PSA_CRYPTO_CONFIG is defined where the library crypto needs are not completely defined by `PSA_WANT_xxx` macros?

> Time-stamp: "2020/09/21 18:07:09 GMT"

Author: > Do we want to support the case when `MBEDTLS_PSA_CRYPTO_CONFIG` is defined where the library crypto needs are not completely defined by `PSA_WANT_xxx` macros?

Yes. A classic crypto feature may need to be enabled either because it's needed internally to implement a PSA crypto feature or because the application needs that classic feature. ``MBEDTLS_PSA_CRYPTO_CONFIG` only determines whether PSA crypto features are deduced from classic crypto options (off) or configured separately (on). If `MBEDTLS_PSA_CRYPTO_CONFIG` is enabled, the PSA crypto configuration (including driver availability) determines a set of classic crypto options that are enabled in addition to those that are enabled explicitly.

> Time-stamp: "2020/09/21 18:07:09 GMT"

Reviewer: ok, thanks for clarifying this.

* If `MBEDTLS_PSA_CRYPTO_CONFIG` is unset, which is the default at least in Mbed TLS 2.x versions, things are as they are today: the PSA subsystem includes generic code unconditionally, and includes support for specific mechanisms conditionally based on the existing `MBEDTLS_xxx_` symbols.
* If `MBEDTLS_PSA_CRYPTO_CONFIG` is set, the necessary software implementations of cryptographic algorithms are included based on both the content of the PSA Crypto configuration file and the Mbed TLS configuration file. For example, the code in `aes.c` is enabled if either `mbedtls/config.h` contains `MBEDTLS_AES_C` or `psa/crypto_config.h` contains `PSA_WANT_KEY_TYPE_AES`.

### PSA Crypto configuration symbols

#### Configuration symbol syntax

A PSA Crypto configuration symbol is a C preprocessor symbol whose name starts with `PSA_WANT_`.

* If the symbol is not defined, the corresponding feature is not included.
* If the symbol is defined to a preprocessor expression with the value `1`, the corresponding feature is included.
* If the symbol is defined with a different value, the behavior is currently undefined and reserved for future use.

#### Configuration symbol usage

> The presence of a symbol `PSA_WANT_xxx` in the Mbed TLS configuration determines whether a feature is available. These symbols should be used in any place that requires conditional compilation based on the availability of a cryptographic mechanism, including:

Reviewer: ```suggestion
The presence of a symbol `PSA_WANT_xxx` in the Mbed TLS configuration determines whether a PSA feature is available. These symbols should be used in any place that requires conditional compilation based on the availability of a cryptographic mechanism, including:
```

> The presence of a symbol `PSA_WANT_xxx` in the Mbed TLS configuration determines whether a feature is available through the PSA API. These symbols should be used in any place that requires conditional compilation based on the availability of a cryptographic mechanism, including:

Reviewer: ```suggestion
The presence of a symbol `PSA_WANT_xxx` in the Mbed TLS configuration determines whether a feature is available through the PSA API. These symbols should be used in any place that requires conditional compilation based on the availability of a cryptographic mechanism through the PSA API, including:
```

### PSA Crypto configuration file

> The PSA crypto configuration file `psa/crypto_config.h` defines a series of symbols of the form `PSA_WANT_xxx` where `xxx` describes the feature that the symbol enables. The symbols are documented in the section [“PSA Crypto configuration symbols”](#psa-crypto-configuration-symbols) below.

Reviewer 2: Minor
```suggestion
The PSA Crypto configuration file `psa/crypto_config.h` defines a series of symbols of the form `PSA_WANT_xxx` where `xxx` describes the feature that the symbol enables. The symbols are documented in the section [“PSA Crypto configuration symbols”](#psa-crypto-configuration-symbols) below.
```

* If the symbol is defined to a preprocessor expression with the value `1`, the corresponding feature is included.
* If the symbol is defined with a different value, the behavior is currently undefined and reserved for future use.

#### Configuration symbol usage

> The presence of a symbol `PSA_WANT_xxx` in the Mbed TLS configuration determines whether a feature is available. These symbols should be used in any place that requires conditional compilation based on the availability of a cryptogrraphic mechanism, including:

Reviewer 2: Minor:
```suggestion
The presence of a symbol `PSA_WANT_xxx` in the Mbed TLS configuration determines whether a feature is available. These symbols should be used in any place that requires conditional compilation based on the availability of a cryptographic mechanism, including:
```

#### Definition of internal inclusion symbols

When `MBEDTLS_PSA_CRYPTO_CONFIG` is set, the header file `mbedtls/config.h` needs to define all the `MBEDTLS_xxx_C` configuration symbols, including the ones deduced from the PSA crypto configuration. It does this by including the new header file **`mbedtls/config_psa.h`**, which defines the `MBEDTLS_PSA_BUILTIN_xxx` symbols and deduces the corresponding `MBEDTLS_xxx_C` (and other) symbols.

> When `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, the configuration of Mbed TLS works as before, and the inclusion of non-PSA code only depends on `MBEDTLS_xxx` symbols defined (or not) in `mbedtls/config.h`. In this case, `mbedtls/config_psa.h` is only needed to build the PSA parts of the library, including `psa/crypto_struct.h`. Therefore, * `psa/crypto_struct.h` needs to include `mbedtls/config_psa.h`.

Reviewer: "Therefore, `psa/crypto_struct.h` needs to include `mbedtls/config_psa.h`." I don't really understand the logical implication here and do we need to go into such details here?

Author: Explanation added. Going into this kind of detail is the point of this section.

> When `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, the configuration of Mbed TLS works as before, and the inclusion of non-PSA code only depends on `MBEDTLS_xxx` symbols defined (or not) in `mbedtls/config.h`. In this case, `mbedtls/config_psa.h` is only needed to build the PSA parts of the library, including `psa/crypto_struct.h` (where the layout of some structures depends on the configuration). Therefore, `psa/crypto_struct.h` (or a header that is included before it) needs to include `mbedtls/config_psa.h`.

Reviewer: I still don't really understand why you put some special emphasis on `psa/crypto_struct.h`. The definitions in the header depends on the configuration of the library thus it includes `mbedtls/config.h` like every .c or .h file in the project with a dependency on the configuration.  Is it the only one in `include/psa` ? No, I can see that `crypto_sizes.h` and `crypto_platform.h` include also `mbedtls/config.h`.

Regarding the case where  `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, one important thing to me that is only visible in the table at the end (if I didn't miss it somewhere else) is that `mbedtls/config_psa.h` still defines `MBEDTLS_PSA_BUILTIN_xxx` symbols but based on `MBEDTLS_xxx_C` symbols not `PSA_WANT_xxx` symbols.

Author: Good point about `crypto_sizes.h`. The point is that it needs to be included in public headers.

Reviewer: > The point is that it needs to be included in public headers.

ok we are hitting the fact that the project's header policy is not well defined here I'd say. It would be better not having to mention that, it should be obvious.

Author: I've reworded this paragraph, in particular to state explicitly that application code needs `PSA_WANT_xxx`. I think this should be explicit since we got it wrong at one point.

`mbedtls/config_psa.h` includes two header files:

* `psa/crypto_config.h` is the user-editable file that defines application requirements. It is only included when `MBEDTLS_PSA_CRYPTO_CONFIG` is set.
* `mbedtls/crypto_drivers.h` is a header file generated by the transpilation of the driver descriptions. It defines `MBEDTLS_PSA_ACCEL_xxx` symbols according to the availability of transparent drivers without fallback.

The following table summarized where symbols are defined depending on the configuration mode.

* (U) indicates a symbol that is defined by the user (application).
* (D) indicates a symbol that is deduced from other symbols by code that ships with Mbed TLS.
* (G) indicates a symbol that is generated from driver descriptions.

------------------------------------------------------------------------------------------------
Symbols                    With `MBEDTLS_PSA_CRYPTO_CONFIG`  Without `MBEDTLS_PSA_CRYPTO_CONFIG`
-------------------------  --------------------------------  -----------------------------------
`MBEDTLS_xxx_C`            `mbedtls/config.h` (U) or         `mbedtls/config.h` (U)
                           `mbedtls/config_psa.h` (D)

> `PSA_WANT_xxx`             `psa/crypto_config.h` (U)         N/A

Author: Without `CONFIG`, `PSA_WANT_xxx` also needs to be defined, for the sake of code that calls the PSA API (TLS code, tests, sample apps). https://github.com/ARMmbed/mbedtls/issues/3670#issuecomment-697966113

`MBEDTLS_PSA_BUILTIN_xxx`  `mbedtls/config_psa.h` (D)        `mbedtls/config_psa.h` (D)

> `MBEDTLS_PSA_ACCEL_xxx`    `mbedtls/crypto_drivers` (G)      N/A

Reviewer: Typo: "mbedtls/crypto_drivers" instead of "mbedtls/crypto_drivers.h"

Reviewer: Otherwise without `MBEDTLS_PSA_CRYPTO_CONFIG` in config_psa.h only `MBEDTLS_PSA_BUILTIN_xxx` macros are deduced not `PSA_WANT_xxx` ones I think.

Author: > Otherwise without MBEDTLS_PSA_CRYPTO_CONFIG in config_psa.h only MBEDTLS_PSA_BUILTIN_xxx macros are deduced not PSA_WANT_xxx ones I think.

No, `PSA_WANT_xxx` is needed in both modes. See https://github.com/ARMmbed/mbedtls/pull/3697#discussion_r493888080 and https://github.com/ARMmbed/mbedtls/issues/3670#issuecomment-697966113 . I've removed the commit that removed it from the table, and will add this and the rationale to the text.

### Architecture of symbol definitions

#### New-style definition of configuration symbols

When `MBEDTLS_PSA_CRYPTO_CONFIG` is set, the header file `mbedtls/config.h` needs to define all the `MBEDTLS_xxx_C` configuration symbols, including the ones deduced from the PSA Crypto configuration. It does this by including the new header file **`mbedtls/config_psa.h`**, which defines the `MBEDTLS_PSA_BUILTIN_xxx` symbols and deduces the corresponding `MBEDTLS_xxx_C` (and other) symbols.

`mbedtls/config_psa.h` includes `psa/crypto_config.h`, the user-editable file that defines application requirements.

#### New-style definition of configuration symbols

> When `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, the configuration of Mbed TLS works as before, and the inclusion of non-PSA code only depends on `MBEDTLS_xxx` symbols defined (or not) in `mbedtls/config.h`. Furthermore, a new header file **`mbedtls/config_psa.h`** deduces PSA configuration symbols (`PSA_WANT_xxx`, `MBEDTLS_PSA_BUILTIN_xxx`) from classic configuration symbols (`MBEDTLS_xxx`).

Author: “A new header file” is correct, why change to “the”?

You objected to “in that case” in an earlier version, and since “that case” is now the whole section it would be more out of place now.

> When `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, the configuration of Mbed TLS works as before, and the inclusion of non-PSA code only depends on `MBEDTLS_xxx` symbols defined (or not) in `mbedtls/config.h`. Furthermore, a new header file **`mbedtls/config_psa.h`** deduces PSA configuration symbols (`PSA_WANT_xxx`, `MBEDTLS_PSA_BUILTIN_xxx`) from classic configuration symbols (`MBEDTLS_xxx`).

Reviewer: About "the" it is because in the previous section config_psa.h is introduced as "the new header file **`mbedtls/config_psa.h`**" thus using "a" now is a bit weird to me, a bit like it is another file. I think there is some kind of a rule that when you first introduce something new you use "a" and after you use "the" thus here I'd rather do it the other way around: "a" in the previous paragraph and "the" here. But I can be wrong.

> When `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, the configuration of Mbed TLS works as before, and the inclusion of non-PSA code only depends on `MBEDTLS_xxx` symbols defined (or not) in `mbedtls/config.h`. Furthermore, a new header file **`mbedtls/config_psa.h`** deduces PSA configuration symbols (`PSA_WANT_xxx`, `MBEDTLS_PSA_BUILTIN_xxx`) from classic configuration symbols (`MBEDTLS_xxx`).

Author: “A” is correct here since this sentence introduces the new file. But “the” is also correct, so I've changed it to “the”.

#### Definition of internal inclusion symbols

When `MBEDTLS_PSA_CRYPTO_CONFIG` is set, the header file `mbedtls/config.h` needs to define all the `MBEDTLS_xxx_C` configuration symbols, including the ones deduced from the PSA Crypto configuration. It does this by including the new header file **`mbedtls/config_psa.h`**, which defines the `MBEDTLS_PSA_BUILTIN_xxx` symbols and deduces the corresponding `MBEDTLS_xxx_C` (and other) symbols.

> When `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, the configuration of Mbed TLS works as before, and the inclusion of non-PSA code only depends on `MBEDTLS_xxx` symbols defined (or not) in `mbedtls/config.h`. In this case, `mbedtls/config_psa.h` is only needed to build the PSA parts of the library and code that uses these parts. Note that `psa/crypto_struct.h` and `psa/crypto_sizes.h` include definitions that depend on the configuration, so `mbedtls/config_psa.h` needs to be included in or before these headers (it isn't enough to include it only from files in `library`).

Reviewer: It would probably be good to add here that when MBEDTLS_PSA_CRYPTO_CONFIG is not set, mbedtls/config_psa.h deduces from MBEDTLS_xxx symbols the `MBEDTLS_PSA_BUILTIN_xxx` and the `PSA_WANT_xxx` symbols.

> When `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, the configuration of Mbed TLS works as before, and the inclusion of non-PSA code only depends on `MBEDTLS_xxx` symbols defined (or not) in `mbedtls/config.h`. In this case, `mbedtls/config_psa.h` deduces PSA configuration symbols (`PSA_WANT_xxx`, `MBEDTLS_PSA_BUILTIN_xxx`) from classic configuration symbols (`MBEDTLS_xxx`). The `PSA_WANT_xxx` definitions in `mbedtls/config_psa.h` are needed not only to build the PSA parts of the library, but also to build code that uses these parts. This includes structure definitions in `psa/crypto_struct.h`, size calculations in `psa/crypto_sizes.h`, and application code that contains code that's specific to a given cryptographic mechanism (which is reflected in Mbed TLS in code under `MBEDTLS_USE_PSA_CRYPTO` and in conditional compilation guards in tests and sample programs).

Reviewer: ```suggestion
When `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, the configuration of Mbed TLS works as before, and the inclusion of non-PSA code only depends on `MBEDTLS_xxx` symbols defined (or not) in `mbedtls/config.h`. Furthermore, `mbedtls/config_psa.h` deduces PSA configuration symbols (`PSA_WANT_xxx`, `MBEDTLS_PSA_BUILTIN_xxx`) from classic configuration symbols (`MBEDTLS_xxx`). The `PSA_WANT_xxx` definitions in `mbedtls/config_psa.h` are needed not only to build the PSA parts of the library, but also to build code that uses these parts. This includes structure definitions in `psa/crypto_struct.h`, size calculations in `psa/crypto_sizes.h`, and application code that contains code that's specific to a given cryptographic mechanism (which is reflected in Mbed TLS in code under `MBEDTLS_USE_PSA_CRYPTO` and in conditional compilation guards in tests and sample programs).
```
"In this case" does not sound well to me here. It sounds better with "furthermore": first sentence for non-PSA code, second for PSA code and PSA-dependent code.

Author: I've reorganized this section a bit.

> When `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, the configuration of Mbed TLS works as before, and the inclusion of non-PSA code only depends on `MBEDTLS_xxx` symbols defined (or not) in `mbedtls/config.h`. In this case, `mbedtls/config_psa.h` deduces PSA configuration symbols (`PSA_WANT_xxx`, `MBEDTLS_PSA_BUILTIN_xxx`) from classic configuration symbols (`MBEDTLS_xxx`). The `PSA_WANT_xxx` definitions in `mbedtls/config_psa.h` are needed not only to build the PSA parts of the library, but also to build code that uses these parts. This includes structure definitions in `psa/crypto_struct.h`, size calculations in `psa/crypto_sizes.h`, and application code that contains code that's specific to a given cryptographic mechanism (which is reflected in Mbed TLS in code under `MBEDTLS_USE_PSA_CRYPTO` and in conditional compilation guards in tests and sample programs).

Reviewer: ```suggestion
When `MBEDTLS_PSA_CRYPTO_CONFIG` is not set, the configuration of Mbed TLS works as before, and the inclusion of non-PSA code only depends on `MBEDTLS_xxx` symbols defined (or not) in `mbedtls/config.h`. In this case, `mbedtls/config_psa.h` deduces PSA configuration symbols (`PSA_WANT_xxx`, `MBEDTLS_PSA_BUILTIN_xxx`) from classic configuration symbols (`MBEDTLS_xxx`). The `PSA_WANT_xxx` definitions in `mbedtls/config_psa.h` are needed not only to build the PSA parts of the library, but also to build code that uses these parts. This includes structure definitions in `psa/crypto_struct.h`, size calculations in `psa/crypto_sizes.h`, and application code containing code specific to a given cryptographic mechanism (which is reflected in code under `MBEDTLS_USE_PSA_CRYPTO` in Mbed TLS and in conditional compilation guards in tests and sample programs).
```

`mbedtls/config_psa.h` includes two header files:

* `psa/crypto_config.h` is the user-editable file that defines application requirements. It is only included when `MBEDTLS_PSA_CRYPTO_CONFIG` is set.

> * `mbedtls/crypto_drivers.h` is a header file generated by the transpilation of the driver descriptions. It defines `MBEDTLS_PSA_ACCEL_xxx` symbols according to the availability of transparent drivers without fallback.

Reviewer 2: What does "transpilation" mean?

Author: It's a fancy word for a compiler whose output is in a language that's generally considered source code. More or less synonymous with preprocessor, but preprocessor implies that the source language includes the target language or something close.
